librz/io/p/io_srec.c

```diff
@@ -0,0 +1,797 @@
+
+S00F000068656C6C6F202020202000003C
+S11F00007C0802A6900100049421FFF07C6C1B787C8C23783C6000003863000026
+S11F001C4BFFFFE5398000007D83637880010014382100107C0803A64E800020E9
+S5030003F9
+S9030000FC
+*/
+
+#include <rz_io.h>
+#include <rz_lib.h>
+
+#define NMAX             100
+#define SREC_PATH_PREFIX "srec://"
+
+// struct RzSREC : holds sparse buffer + its own fd, for internal management
+typedef struct {
+	st32 fd;
+	RzBuffer *rbuf;
+} RzSREC;
+
+static st32 fw04b(FILE *fd, ut16 eaddr);
+static st32 fwblock(FILE *fd, ut8 *b, ut32 start_addr, ut16 size);
+
+static st32 __write(RzIO *io, RzIODesc *fd, const ut8 *buf, st32 count) {
+	const char *pathname;
+	FILE *out;
+	RzSREC *rih;
+	RzBufferSparse *rbs;
+	RzListIter *iter;
+
+	if (!fd || !fd->data || (fd->perm & RZ_PERM_W) == 0 || count <= 0) {
+		return -1;
+	}
+	rih = fd->data;
+	pathname = fd->name + 7;
+	out = rz_sys_fopen(pathname, "w");
+	if (!out) {
+		eprintf("Cannot open '%s' for writing\n", pathname);
+		return -1;
+	}
+	/* mem write */
+	if (rz_buf_write_at(rih->rbuf, io->off, buf, count) != count) {
+		eprintf("srec:write(): sparse write failed\n");
+		fclose(out);
+		return -1;
+	}
+	rz_buf_seek(rih->rbuf, count, RZ_BUF_CUR);
+
+	/* disk write : process each sparse chunk */
+	// TODO : sort addresses + check overlap?
+	RzList *nonempty = rz_buf_nonempty_list(rih->rbuf);
+	rz_list_foreach (nonempty, iter, rbs) {
+		ut16 addl0 = rbs->from & 0xffff;
+		ut16 addh0 = rbs->from >> 16;
+		ut16 addh1 = rbs->to >> 16;
+		ut16 tsiz = 0;
+		if (rbs->size == 0) {
+			continue;
+		}
+
+		if (addh0 != addh1) {
+			// we cross a 64k boundary, so write in two steps
+			if (fw04b(out, addh0) < 0) {
+				eprintf("srec:write: file error\n");
+				rz_list_free(nonempty);
+				fclose(out);
+				return -1;
+			}
+			tsiz = -addl0;
+			addl0 = 0;
+			if (fwblock(out, rbs->data, rbs->from, tsiz)) {
+				eprintf("srec:fwblock error\n");
+				rz_list_free(nonempty);
+				fclose(out);
+				return -1;
+			}
+		}
+		if (fw04b(out, addh1) < 0) {
+			eprintf("srec:write: file error\n");
+			rz_list_free(nonempty);
+			fclose(out);
+			return -1;
+		}
+		if (fwblock(out, rbs->data + tsiz, (addh1 << 16) | addl0, rbs->size - tsiz)) {
+			eprintf("srec:fwblock error\n");
+			rz_list_free(nonempty);
+			fclose(out);
+			return -1;
+		}
+	} // list_foreach
+
+	rz_list_free(nonempty);
+	fprintf(out, ":00000001FF\n");
+	fclose(out);
+	out = NULL;
+	return 0;
+}
+
+// write contiguous block of data to file; ret 0 if ok
+// max 65535 bytes;
+static st32 fwblock(FILE *fd, ut8 *b, ut32 start_addr, ut16 size) {
+	ut8 cks;
+	char linebuf[80];
+	ut16 last_addr;
+	st32 j;
+	ut32 i; // has to be bigger than size !
+
+	if (size < 1 || !fd || !b) {
+		return -1;
+	}
+
+	for (i = 0; (i + 0x10) < size; i += 0x10) {
+		cks = 0x10;
+		cks += (i + start_addr) >> 8;
+		cks += (i + start_addr);
+		for (j = 0; j < 0x10; j++) {
+			cks += b[j];
+		}
+		cks = 0 - cks;
+		if (fprintf(fd, ":10%04x00%02x%02x%02x%02x%02x%02x%02x"
```

Comment:
Wait, that's not correct, is it? `:10` is IHEX, what you want is `S113` for "data record w. 16bit address, byte count 19"